### PR TITLE
"Deadlock Avoidance" PR without stylistic changes

### PR DESCRIPTION
**Pull request by watfordgnf**
This is PR #71 without any stylistic changes. This should help with reviewing the details of it without getting bogged down following whitespace or naming changes.

### CODE REVIEW

**Comment by watfordgnf** [comment]
Squashed commits.

File: STAN.CLIENT/StanConnection.cs [hunk 1]
```diff
@@ -469,61 +408,46 @@ public IStanSubscription Subscribe(string subject, EventHandler<StanMsgHandlerAr
...
-                throw new ArgumentNullException("handler");
+                throw new ArgumentNullException(nameof(handler));
 
             return subscribe(subject, qgroup, handler, options);
         }
 
         public void Close()
         {
-            Msg reply = null;
+            if (IsClosed)
```
Comment (watfordgnf) [inline STAN.CLIENT/StanConnection.cs]: @harold-toledo the definition of IsClosed would allow this to be called multiple times if LeaveOpen is true.

Comment (harold-toledo) [reply STAN.CLIENT/StanConnection.cs]: I do not see how LeaveOpen could potentially affect this call, and the fact that we could potentially call it multiple times should not cause any issues.

**Comment by watfordgnf** [comment]
@ColinSullivan1 these look to be related to the different semantics in the Close path with this PR; from what I can tell at a high level:
1. Closing a subscription no longer removes the connection.
2. Closing a subscription may result in a close request being made more than once.

**Comment by ColinSullivan1** [comment]
@watfordgnf, I think this is really useful as a review tool.  However, how do you feel about using the changes in PR #71?  I'm OK moving toward code [that is more idiomatic to C#](https://github.com/nats-io/csharp-nats#about-the-code-and-contributing).  The author contacted me and I understand it has been running in production and has been well tested.

**Pull request closed without merging**